api.go

```diff
@@ -200,7 +202,217 @@ func (r *Router) RouterCallImpl(ctx context.Context,
+	if opts.Timeout == 0 {
+		opts.Timeout = CallTimeoutMin
+	}
+
+	timeout := opts.Timeout
+	timeStart := time.Now()
+
+	refID := r.refID.Load()
+	r.refID.Add(1)
+
+	defer r.callStorageUnref(refID)
+
+	req := tarantool.NewCallRequest("vshard.storage._call")
+	req = req.Context(ctx)
+
+	// ref stage: send
+
+	req = req.Args([]interface{}{
+		"storage_ref",
+		refID,
+		timeout,
+	})
+
+	cases := r.getMapCallSelectCases(req)
+
+	// ref stage collect
+
+	totalBucketCount := uint16(0)
+
+	for i := 0; i < len(cases); i++ {
+		_, resp, ok := reflect.Select(cases)
+		if !ok {
+			return nil, errors.New("failed to select a case from metaDatas")
+		}
+
+		metaData, ok := resp.Interface().(*MapCallMetaData)
+		if !ok {
+			return nil, errors.New("the type of received data is not *MapCallMetaData")
+		}
+
+		respData, err, future := metaData.Resp, metaData.Error, metaData.future
+		if err != nil {
+			return nil, err
+		}
+
+		if respData[0] == nil {
+			vshardErr := &StorageCallAssertError{}
+
+			err = mapstructure.Decode(respData[1], vshardErr)
+			if err != nil {
+				return nil, err
+			}
+
+			return nil, vshardErr
+		}
+
+		var bucketCount uint16
+		err = future.GetTyped(&[]interface{}{&bucketCount})
+		if err != nil {
+			return nil, err
+		}
+
+		totalBucketCount += bucketCount
+	}
+
+	if int32(totalBucketCount) != r.knownBucketCount.Load() {
+		return nil, fmt.Errorf("unknown bucket counts %d", totalBucketCount)
+	}
+
+	// map stage: send
+
+	req = req.Args([]interface{}{"storage_map", refID, fnc, args})
+
+	cases = r.getMapCallSelectCases(req)
+
+	// map stage: collect
+
+	idToResult := make(map[uuid.UUID]interface{})
+
+	for i := 0; i < len(cases); i++ {
+		_, resp, ok := reflect.Select(cases)
+		if !ok {
+			return nil, errors.New("failed to select a case from metaDatas")
+		}
+
+		metaData, ok := resp.Interface().(*MapCallMetaData)
+		if !ok {
+			return nil, errors.New("the type of received data is not *MapCallMetaData")
+		}
+
+		respData, err, future := metaData.Resp, metaData.Error, metaData.future
+		if err != nil {
+			return nil, err
+		}
+
+		if len(respData) != 2 {
+			err = fmt.Errorf("invalid length of response data: must be = 2, current: %d", len(respData))
+
+			return nil, err
+		}
+
+		if respData[0] == nil {
+			vshardErr := &StorageCallAssertError{}
+
+			err = mapstructure.Decode(respData[1], vshardErr)
+			if err != nil {
+				return nil, err
+			}
+
+			return nil, vshardErr
+		}
+
+		isVShardRespOk := false
+		err = future.GetTyped(&[]interface{}{&isVShardRespOk})
+		if err != nil {
+			return nil, err
+		}
+
+		if !isVShardRespOk { // error
+			errorResp := &StorageCallAssertError{}
+
+			err = future.GetTyped(&[]interface{}{&isVShardRespOk, errorResp})
+			if err != nil {
+				err = fmt.Errorf("cant get typed vshard err with err: %s", err)
+			}
+
+			return nil, err
+		}
+
+		id := metaData.replicaset.info.UUID
+		idToResult[id] = respData[1]
+	}
+
+	r.metrics().RequestDuration(time.Since(timeStart), true)
```

Comment:
i think we need one more label for this metric, like bool flag for map and non-map requests

Reply:
make flag mapReduce